src/filetype-plugin.ts

```diff
@@ -0,0 +1,88 @@
+
+    function add_category_extensions(category: Category, exts: string[]) {
+        for (let ext of exts) {
+            if (ext[0] === '.') {
+                ext = ext.substring(1);
+            }
+
+            max_extension_length = Math.max(max_extension_length, ext.length);
+            extensions[ext] = category;
+        }
+    }
+
+    // Broad-stroke categorization based on 'mime-db' package
+    for (const [mimetype, details] of Object.entries(mime_db)) {
+        if ('extensions' in details) {
+            if (mimetype.startsWith('image/')) {
+                add_category_extensions(Category.IMAGE, details.extensions);
+            } else if (mimetype.startsWith('audio/')) {
+                add_category_extensions(Category.AUDIO, details.extensions);
+            } else if (mimetype.startsWith('video/')) {
+                add_category_extensions(Category.VIDEO, details.extensions);
+            } else if (mimetype.startsWith('text/')) {
+                add_category_extensions(Category.TEXT, details.extensions);
+            }
+        }
+    }
+
+    // Archives, scraped from https://en.wikipedia.org/wiki/List_of_archive_formats
+    add_category_extensions(
+        Category.ARCHIVE,
+        [
+            '7z', 'F', 'LBR', 'Z', 'a', 'aar', 'ace', 'afa', 'alz', 'apk', 'ar', 'arc', 'arc', 'arj', 'ark',
+            'b1', 'b6z', 'ba', 'bh', 'br', 'bz2', 'cab', 'car', 'cdx', 'cfs', 'cpio', 'cpt', 'dar', 'dd',
+            'dgc', 'ear', 'gca', 'genozip', 'genozip', 'gz', 'ha', 'hki', 'ice', 'iso', 'jar', 'kgb', 'lbr',
+            'lha', 'lz', 'lz4', 'lzh', 'lzma', 'lzo', 'lzx', 'mar', 'pak', 'paq6', 'paq7', 'paq8', 'partimg',
+            'pea', 'phar', 'pim', 'pit', 'qda', 'rar', 'rk', 'rz', 's7z', 'sbx', 'sda', 'sea', 'sen', 'sfark',
+            'sfx', 'shar', 'shk', 'sit', 'sitx', 'sqx', 'sz', 'tar', 'tar.Z', 'tar.bz2', 'tar.gz', 'tar.lz',
+            'tar.xz', 'tar.zst', 'tbz2', 'tgz', 'tlz', 'txz', 'uc', 'uc0', 'uc2', 'uca', 'ucn', 'ue2', 'uha',
+            'ur2', 'war', 'wim', 'xar', 'xp3', 'xz', 'yz1', 'z', 'zip', 'zipx', 'zoo', 'zpaq', 'zst', 'zz'
+        ]
+    );
+
+    // Special treatment for programming languages based on GitHub's database
+    for (const lang of Object.values(language_map)) {
+        if (lang.type === 'programming' && 'extensions' in lang) {
+            add_category_extensions(Category.CODE, lang.extensions);
+        }
+    }
+
+    return { categories, extensions, max_extension_length };
+}
+
+export class FileTypePlugin implements Plugin {
+    name = 'cockpit-files-filetype-plugin';
+
+    setup(build: PluginBuild) {
+        build.onResolve({ filter: /^\.\/filetype-data$/ }, args => ({
+            path: args.path,
+            namespace: 'cockpit-files-filetype-plugin',
+        }));
+
+        build.onLoad({ filter: /.*/, namespace: 'cockpit-files-filetype-plugin' }, () => ({
+            contents: JSON.stringify(create_filetype_data()),
+            loader: 'json',
```

Comment:
So the plugin just exports the created JSON, super interesting :)

Reply:
Yup.  This esbuild plugin stuff is actually pretty straightforward...